protocols/kad/src/record.rs

```diff
@@ -122,6 +122,21 @@ impl Hash for ProviderRecord {
     }
 }
 
+impl PartialEq for ProviderRecord {
+    fn eq(&self, other: &Self) -> bool {
+        // In the [`Hash`] implementation of [`ProviderRecord`] all but the
+        // `key` and `provider` fields are ignored. The following property has
+        // to hold for all inputs: `k1 == k2 ⇒ hash(k1) == hash(k2)`. Thereby
+        // the [`PartialEq`] implementation has to ignore `key` and `provider`
+        // as well.
```

Comment:
I don't see how that follows since `k1 == k2 ⇒ hash(k1) == hash(k2)` certainly holds with the derived `PartialEq`/`Eq`. The required implication only goes one way. Am I overlooking something?

Reply:
In my eyes this should go both ways, thus I should have written `<=>`. I would expect two equal structs to derive to the same hash, and I would expect two equal hashes to be derived from two equal structs (ignoring the probability of hash collisions).

I have pushed d248c2934e14bba9059d1f8887ab7b493801996b. Do you agree with my intuition?

Reply:
> [..] and I would expect two equal hashes to be derived from two equal structs

Since these are not cryptographic hashes, that is not a reasonable expectation. That hash collisions are totally fine is part of  `std::hash::Hash`es semantics for use with hash-based data structures.